joomla/joomla-cms
# [4.0] Namespace JHtml

This namespaces the JHtml helpers as agreed by Production. Fixes #28249

- Also fixes a bug in the `error_login.php` file of atum that I discovered whilst doing this where we could double register a html service

### Summary of Changes
Namespaces JHtml classes. Note the two classes that are new for J4 (workflowstages and draggable do not have namespace mappings created as they are new for j4)

### Testing Instructions
Ensure CMS works as before in a variety of frontend and backend views

If you are technical review class names make sense - as they aren’t tied to the name being called they can be made longer without any affect

### Documentation Changes Required
Yes. If you are overriding JHtml methods you can no longer use `\Joomla\CMS\HTML\HTMLHelper::register()` you now must instead use

`\Joomla\CMS\HTML\HTMLHelper::getServiceRegistry()->register('jquery', 'MyCustomClass', true);`

The last parameter `$replace` is the important part :) Note that this means you can only override the entire Helper. You can't overload a specific function after this change. In the two cases where people regularly do this in the ecosystem (bootstrap and jquery) this isn't really an issue:
- In JQuery there are only two methods - the second a helper for ajax and it's a minor thing to have to duplicate and I don't see any real issues forthcoming as jQuery is now fairly static
- In Bootstrap whilst there are lots of methods they output html specific to the bootstrap version so you'd need to have overridden them anyhow
